drivers/serial/uart_ns16550.c

```diff
@@ -733,10 +821,13 @@ static void uart_ns16550_isr(void *arg)
 {
 	struct device *dev = arg;
 	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
+	k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
```

Comment:
This one is causing `test_uart_fifo_fill` to fail

Reply:
Is it spinning with the lock held waiting on a full fifo to drain?  I bet it's spinning with the lock held waiting on a full fifo to drain.

Reply:
For clarity, because my joke wasn't as obvious (or funny) as it seemed when I wrote it: that would deadlock the ISR, which would be needed to wake up a reader to drain the fifo.

Reply:
> For clarity, because my joke wasn't as obvious (or funny) as it seemed ...

Fairly speaking, concerns like this were the reason why I popped up in this (and/or related) ticket. To be more precise, my concern that even seemingly working changes (e.g. tested on real hw) may still adversely affect qemu, which doesn't have too faithful 16550 emulation.

Reply:
@nashif  Thanks for helping narrow this down! How did you figure it out? 
@andyross lol thanks. the `test_uart_fifo_fill` passes if I remove the locking here, cool. So now I need to figure out how to add it back without breaking things...

Reply:
As @andyross said, this lock can be removed. The callback are (supposedly) using the `poll_in()`, `poll_out()`, `fifo_read()` (etc., etc.) to read or write data. If a lock is held in `_isr()`, the read/write functions are waiting for lock to be released (and can't since they are called inside callback), resulting in deadlock.